File: common/src/main/java/juuxel/adorn/fluid/FluidAmountPredicate.java
package juuxel.adorn.fluid;

import net.minecraft.core.component.DataComponentPatch;
import net.minecraft.world.level.material.Fluids;

public interface FluidAmountPredicate {
    HasFluidAmount getUpperBound();

    boolean test(long amount, FluidUnit unit);

    static FluidAmountPredicate exactly(long amount, FluidUnit unit) {
        return new FluidAmountPredicate() {
            private final FluidVolume upperBound = new FluidVolume(Fluids.EMPTY, amount, DataComponentPatch.EMPTY, unit);

            @Override
            public HasFluidAmount getUpperBound() {
                return upperBound;
            }

            @Override
            public boolean test(long amount, FluidUnit unit) {
                return FluidUnit.compareVolumes(amount, unit, upperBound.getAmount(), upperBound.getUnit()) == 0;
            }
        };
    }

    static FluidAmountPredicate atMost(long max, FluidUnit unit) {
        return new FluidAmountPredicate() {
            private final FluidVolume upperBound = new FluidVolume(Fluids.EMPTY, max, DataComponentPatch.EMPTY, unit);

            @Override
            public HasFluidAmount getUpperBound() {
                return upperBound;
            }

            @Override
            public boolean test(long amount, FluidUnit unit) {
                return FluidUnit.compareVolumes(amount, unit, upperBound.getAmount(), upperBound.getUnit()) <= 0;
            }
        };
    }
}

File: common/src/main/kotlin/juuxel/adorn/fluid/FluidAmountPredicate.kt
package juuxel.adorn.fluid

import net.minecraft.fluid.Fluids

interface FluidAmountPredicate {
    val upperBound: HasFluidAmount

    fun test(amount: Long, unit: FluidUnit): Boolean

    companion object {
        fun exactly(amount: Long, unit: FluidUnit): FluidAmountPredicate {
            return object : FluidAmountPredicate {
                override val upperBound = FluidVolume(Fluids.EMPTY, amount, null, unit)

                override fun test(amount: Long, unit: FluidUnit): Boolean =
                    FluidUnit.compareVolumes(amount, unit, upperBound.amount, upperBound.unit) == 0
            }
        }

        fun atMost(max: Long, unit: FluidUnit): FluidAmountPredicate {
            return object : FluidAmountPredicate {
                override val upperBound = FluidVolume(Fluids.EMPTY, max, null, unit)

                override fun test(amount: Long, unit: FluidUnit): Boolean =
                    FluidUnit.compareVolumes(amount, unit, upperBound.amount, upperBound.unit) == 0
            }
        }
    }
}
